blockchain/chain_sync/src/tipset_syncer.rs

```diff
@@ -0,0 +1,1741 @@
+            // Tipset we a synchronizing toward, stop.
+            break;
+        }
+        // Attempt to load the parent tipset from local store
+        if let Ok(tipset) = chain_store.tipset_from_keys(oldest_parent.parents()).await {
+            parent_blocks.extend_from_slice(tipset.cids());
+            parent_tipsets.push(tipset);
+            continue;
+        }
+
+        // TODO: Tweak request window when socket frame is tested
+        let epoch_diff = oldest_parent.epoch() - current_head.epoch();
+        let window = min(epoch_diff, MAX_TIPSETS_TO_REQUEST as i64);
+        trace!(
+            "ChainExchange for TipSet range: {} -> {}",
+            oldest_parent.epoch(),
+            oldest_parent.epoch() - window
+        );
+        let network_tipsets = network
+            .chain_exchange_headers(None, oldest_parent.parents(), window as u64)
+            .await
+            .map_err(|err| TipsetRangeSyncerError::NetworkTipsetQueryFailed(err.to_string()))?;
+        trace!(
+            "Got tipsets: Height: {}, Len: {}",
+            network_tipsets[0].epoch(),
+            network_tipsets.len()
+        );
+
+        for tipset in network_tipsets {
+            // Break if have already traversed the entire tipset range
+            if tipset.epoch() < current_head.epoch() {
+                break 'sync;
+            }
+            validate_tipset_against_cache(bad_block_cache.clone(), &tipset.key(), &parent_blocks)
+                .await?;
+            parent_blocks.extend_from_slice(tipset.cids());
+            tracker.write().await.set_epoch(tipset.epoch());
+            parent_tipsets.push(tipset);
+        }
+    }
+    // Unwrapping is safe here because we assume that the tipset
+    // vector was initialized with a tipset that will not be removed
+    let oldest_tipset = parent_tipsets.last().unwrap().clone();
+    // Determine if the local chain was a fork.
+    // If it was, then sync the fork tipset range by iteratively walking back
+    // from the oldest tipset synced until we find a common ancestor
+    if oldest_tipset.parents() != current_head.parents() {
+        const FORK_LENGTH_THRESHOLD: u64 = 900;
+        let fork_tipsets = network
+            .chain_exchange_headers(None, oldest_tipset.parents(), FORK_LENGTH_THRESHOLD)
+            .await
+            .map_err(|err| TipsetRangeSyncerError::NetworkTipsetQueryFailed(err.to_string()))?;
+        let mut potential_common_ancestor =
+            chain_store.tipset_from_keys(current_head.parents()).await?;
+        let fork_length = 1;
+        for (i, tipset) in fork_tipsets.iter().enumerate() {
+            if tipset.epoch() == 0 {
+                return Err(TipsetRangeSyncerError::ForkAtGenesisBlock(format!(
+                    "{:?}",
+                    oldest_tipset.cids()
+                )));
+            }
+            if potential_common_ancestor == *tipset {
+                // TODO: Iterate over the blocks in the forks tipsets,
```

Comment:
Did we miss it or in a later PR?

Reply:
Is this really `TODO` ? And if so, is it essential functionality? Or will it ever result in a consensus fault?

Reply:
It is not essential functionality and our current implementation assumes that blocks received over chain exchange requests are valid. So, I'm going to remove this TODO and leave as is.